File: Week_01/java/src/_696_CountBinarySubstrings.java
/**
 * 第一遍：2020/08/10周一 ✅
 * 第二遍：2020/08/10周二 ✅
 * 第三遍：2020/07/18周六
 * 第三遍：2020/07/08周四
 * 简单题没想出来啊，菜。。
 */
class _696_CountBinarySubstrings {
    public int countBinarySubstrings(String s) {
        if (s == null) return 0;
        int ret = 0;
        int pre = 0;
        int curr = 1;
        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                curr++;
            } else {
                ret += Math.min(curr, pre);
                pre = curr;
                curr = 1;
            }
        }
        return ret + Math.min(curr, pre);
    }
}

File: Week_01/kotlin/src/_696_CountBinarySubstrings.kt
fun countBinarySubstrings(s: String): Int {
    var pre = 0 //前一段0或1的连续子串的长度
    var curr = 1 //当前0或1的连续子串的长度
    var ret = 0 // 符合条件的子字符串的数量
    for (i in 1 until s.length) {
        if (s[i] == s[i - 1]) {
            curr++
        } else {
            ret += minOf(pre, curr)
            pre = curr
            curr = 1
        }
    }
    return ret + minOf(pre, curr)
}